### PR TITLE
Pip: Fallback to obtaining package meta data locally

### PR DESCRIPTION
**Pull request by fviernau**
See the individual commit messages for the details.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 1]
```diff
@@ -424,8 +424,8 @@ class Pip(
 
         // Example license classifier:
         // "License :: OSI Approved :: GNU Library or Lesser General Public License (LGPL)"
-        pkgInfo["classifiers"]?.mapNotNullTo(declaredLicenses) {
-            val components = it.textValue().split(" :: ")
+        pkgInfo["classifiers"]?.mapNotNullTo(declaredLicenses) { classifier ->
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pip.kt]: Commit message: param**e**ter

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 2]
```diff
@@ -424,14 +424,17 @@ class Pip(
...
-        }
+        pkgInfo["classifiers"]?.mapNotNullTo(declaredLicenses) { getLicenseFromClassifier(it.textValue()) }
 
         return declaredLicenses
     }
 
+    private fun getLicenseFromClassifier(classifier: String): String? {
+        val components = classifier.split(" :: ")
+
+        return components.takeIf { it.first() == "License" }?.last()?.takeUnless { it == "OSI Approved" }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pip.kt:442]: I think you could as well eliminate the `components` variable and make this an expression function if you want.

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 3]
```diff
@@ -624,8 +624,10 @@ class Pip(
     }
 }
 
-private fun Package.enrichWith(other: Package): Package =
-    Package(
+private fun Package.enrichWith(other: Package?): Package {
+    if (other == null) return this
+
+    return Package(
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pip.kt:720]: You could keep this an expression function by using:
```
other?.let { Package(...) }
```

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 4]
```diff
@@ -238,6 +238,9 @@ class Pip(
         // List all packages installed locally in the virtualenv.
         val pipdeptree = runInVirtualEnv(virtualEnvDir, workingDir, "pipdeptree", "-l", "--json-tree")
 
+        // Get the locally available meta data for all installed packages as a fallback.
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pip.kt:241]: Commit message: "relatively", "figure out the **declared** license"

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 5]
```diff
@@ -622,6 +628,89 @@ class Pip(
...
+
+                if (value.isNotEmpty()) {
+                    map.getOrPut(key, { mutableListOf() }) += value
+                }
+
+                previousKey = key
+            } else {
+                if (previousKey == null) return@forEach
+
+                map.getOrPut(previousKey!!, { mutableListOf() }) += line.trim()
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pip.kt:691]: Is the `!!` necessary here?

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Pip.kt]: Yes, I however re-wrote it a bit so that it is not necessary anymore.

**Pull request merged**